### PR TITLE
i2c: adding a recovery function at driver init for nrf5

### PR DESCRIPTION
**Pull request by GiulianoFranchetto**
A I2C bus can be stuck if a communication was active during
a system reset (from a crash for example). To ask all the
slaves to release the line, the master must send at least
9 clocks.
The recovery function should be called when init'ing the
driver, as it takes very little time.

Signed-off-by: Giuliano Franchetto <giuliano.franchetto@intellinium.com>

### CODE REVIEW

File: drivers/i2c/i2c_nrf5.c [hunk 1]
```diff
@@ -289,6 +309,9 @@ static int i2c_nrf5_init(struct device *dev)
 
 	twi->ENABLE = TWI_ENABLE_ENABLE_Disabled;
 
+	status = i2c_nrf5_recovery(dev);
```
Comment (lpereira) [inline drivers/i2c/i2c_nrf5.c]: Can this be called only when the bus is stuck (SCL high and SDA low)?

File: drivers/i2c/i2c_nrf5.c [hunk 2]
```diff
@@ -272,6 +272,26 @@ static void i2c_nrf5_isr(void *arg)
 	}
 }
 
+static int i2c_nrf5_recovery(struct device *dev)
```
Comment (anangl) [inline drivers/i2c/i2c_nrf5.c]: This is not a proper I2C recovery procedure. You should start with setting SCL high, then generate clock pulses (1-0-1) on this line until SDA goes high (at this point the master can control SDA again, so there is no need to generate more pulses). There should be maximum 9 such clock pulses (this is the worst-case scenario, when a slave needs to send 8 bits of data and get the ACK/NAK bit). After getting SDA high you need to generate a stop condition, i.e. make a 0-1 transition on  SDA while keeping SCL high, to finally finish this erroneous transaction. The whole thing is quite nice described for instance in [this application note](http://www.analog.com/media/en/technical-documentation/application-notes/54305147357414AN686_0.pdf).

**Review by anangl — CHANGES_REQUESTED** [review]

**Comment by nashif** [comment]
any updates on this? Still relevant?

**Comment by galak** [comment]
Since no response to @nashif last query, close this for now.

**Pull request closed without merging**